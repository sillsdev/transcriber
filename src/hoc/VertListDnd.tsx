import { PropsWithChildren, ReactNode, useState } from 'react';
import {
  DragDropContext,
  Droppable,
  Draggable,
  DropResult,
} from '@hello-pangea/dnd';
import { List, ListItem } from '@mui/material';

export interface DropProp {
  id: string;
  content: ReactNode;
}

// fake data generator
const getItems = (count: number) =>
  Array.from({ length: count }, (_v, k) => k).map((k) => ({
    id: `${k}`,
    content: `item ${k}`,
  }));

// convert children to ItemProps
const cvtChildren = (children?: ReactNode): DropProp[] | undefined => {
  if (Array.isArray(children)) {
    return children.map((child, index) => ({
      id: `${index}`,
      content: child,
    }));
  }
  if (children) {
    return [{ id: '0', content: children }];
  }
};

export interface OnDropProps {
  results: string[];
  oldIndex: number;
  newIndex: number;
}

export interface VertListDndProps extends PropsWithChildren {
  data?: DropProp[];
  dragHandle?: boolean;
  onDrop?: (props: OnDropProps) => void;
}

export const VertListDnd = ({
  data,
  onDrop,
  dragHandle,
  children,
}: VertListDndProps) => {
  const [items, setItems] = useState<DropProp[]>(
    data ?? cvtChildren(children) ?? getItems(10)
  );

  // a little function to help us with reordering the result
  const reorder = (list: DropProp[], startIndex: number, endIndex: number) => {
    const result = Array.from(list);
    const [removed] = result.splice(startIndex, 1);
    result.splice(endIndex, 0, removed);
    return result;
  };

  const onDragEnd = (result: DropResult) => {
    // dropped outside the list
    if (!result.destination) {
      return;
    }
    const newItems = reorder(
      items,
      result.source.index,
      result.destination.index
    );
    setItems(newItems);
    if (onDrop) {
      onDrop({
        results: newItems.map((item) => item.id),
        oldIndex: result.source.index,
        newIndex: result.destination.index,
      });
    }
  };

  // Normally you would want to split things out into separate components.
  // But in this example everything is just done in one place for simplicity
  return (
    <DragDropContext onDragEnd={onDragEnd}>
      <Droppable droppableId="droppable">
        {(provided, snapshot) => (
          <List
            {...provided.droppableProps}
            ref={provided.innerRef}
            sx={{
              bgcolor: snapshot.isDraggingOver ? 'secondary.light' : 'white',
              p: 1,
            }}
          >
            {items.map((item, index) => (
              <Draggable key={item.id} draggableId={item.id} index={index}>
                {(provided, snapshot) => (
                  <ListItem
                    ref={provided.innerRef}
                    {...provided.draggableProps}
                    {...provided.dragHandleProps}
                    sx={{
                      bgcolor: snapshot.isDragging
                        ? 'primary.light'
                        : dragHandle
                        ? 'transparent'
                        : 'lightgrey',
                      mb: 1,
                    }}
                  >
                    {item.content}
                  </ListItem>
                )}
              </Draggable>
            ))}
            {provided.placeholder}
          </List>
        )}
      </Droppable>
    </DragDropContext>
  );
};
