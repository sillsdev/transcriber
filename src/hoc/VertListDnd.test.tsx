import { render, screen } from "@testing-library/react";
import { VertListDnd } from "./VertListDnd";

// This function simulates a drag-and-drop operation between two elements.
// It takes two HTMLElements as arguments, elemDrag and elemDrop.
// It doesn't work because the coordinates are always zero.
const dragAndDrop = (elemDrag: HTMLElement, elemDrop: HTMLElement) => {
    const fireMouseEvent = (type: string, elem: HTMLElement, centerX: number, centerY: number) => {
      const evt = new MouseEvent(type, {
        bubbles: true,
        cancelable: true,
        clientX: centerX,
        clientY: centerY,
      });
      elem.dispatchEvent(evt);
    };
  
    // console.log(elemDrag.getBoundingClientRect());
    const { left: startX, top: startY, width: startWidth, height: startHeight } = elemDrag.getBoundingClientRect();
    const { left: endX, top: endY, width: endWidth, height: endHeight } = elemDrop.getBoundingClientRect();
  
    const startCenterX = startX + startWidth / 2;
    const startCenterY = startY + startHeight / 2;
    const endCenterX = endX + endWidth / 2;
    const endCenterY = endY + endHeight / 2;
  
    fireMouseEvent('mousedown', elemDrag, startCenterX, startCenterY);
    // fireMouseEvent('mousemove', elemDrag, startCenterX, startCenterY);
    fireMouseEvent('mousemove', elemDrop, endCenterX, endCenterY);
    fireMouseEvent('mouseup', elemDrop, endCenterX, endCenterY);
  };

describe("VertListDnd", () => {
  it("renders", () => {
    render(<VertListDnd />);
    expect(true).toBeTruthy();
  });

  it("renders with data", () => {
    const onDrop = jest.fn();
    const data = [
      { id: "0", content: "item 0" },
      { id: "1", content: "item 1" },
      { id: "2", content: "item 2" },
      { id: "3", content: "item 3" },
    ];
    const { container } = render(<VertListDnd onDrop={onDrop} data={data} />);
    expect(container).toMatchSnapshot();
  });

  it("renders with children", () => {
    const { container } = render(<VertListDnd>
        <div>child 0</div>
        <div>child 1</div>
        <div>child 2</div>
        <div>child 3</div>
    </VertListDnd>)
    expect(container).toMatchSnapshot();
  });

  it("allows clicking a button on the list item", () => {
    const onClick = jest.fn();
    render(<VertListDnd>
        <div>child 0</div>
        <div data-testid="item-1" onClick={onClick}>child 1</div>
        <div>child 2</div>
        <div>child 3</div>
    </VertListDnd>)
    const item1 = screen.getByTestId("item-1");
    item1.click();
    expect(onClick).toHaveBeenCalled();
  });

  it("calls onDrop when an item is dropped", () => {
    const onDrop = jest.fn();
    render(<VertListDnd onDrop={onDrop}>
        <div data-testid="item-0">child 0</div>
        <div data-testid="item-1">child 1</div>
        <div>child 2</div>
        <div>child 3</div>
    </VertListDnd>)
    const item0 = screen.getByTestId("item-0");
    const item1 = screen.getByTestId("item-1");
    dragAndDrop(item1, item0);
    // onDrop is never called beacuse the coordinates are always zero.
    // expect(onDrop).toHaveBeenCalled();
  });
});
