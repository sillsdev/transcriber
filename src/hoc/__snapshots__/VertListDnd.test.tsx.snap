// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VertListDnd renders with children 1`] = `
<div>
  <ul
    class="MuiList-root MuiList-padding css-134cfso-MuiList-root"
    data-rfd-droppable-context-id=":ra:"
    data-rfd-droppable-id="droppable"
  >
    <li
      aria-describedby="rfd-hidden-text-:ra:-hidden-text-:rb:"
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-6leons-MuiListItem-root"
      data-rfd-drag-handle-context-id=":ra:"
      data-rfd-drag-handle-draggable-id="0"
      data-rfd-draggable-context-id=":ra:"
      data-rfd-draggable-id="0"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <div>
        child 0
      </div>
    </li>
    <li
      aria-describedby="rfd-hidden-text-:ra:-hidden-text-:rb:"
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-6leons-MuiListItem-root"
      data-rfd-drag-handle-context-id=":ra:"
      data-rfd-drag-handle-draggable-id="1"
      data-rfd-draggable-context-id=":ra:"
      data-rfd-draggable-id="1"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <div>
        child 1
      </div>
    </li>
    <li
      aria-describedby="rfd-hidden-text-:ra:-hidden-text-:rb:"
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-6leons-MuiListItem-root"
      data-rfd-drag-handle-context-id=":ra:"
      data-rfd-drag-handle-draggable-id="2"
      data-rfd-draggable-context-id=":ra:"
      data-rfd-draggable-id="2"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <div>
        child 2
      </div>
    </li>
    <li
      aria-describedby="rfd-hidden-text-:ra:-hidden-text-:rb:"
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-6leons-MuiListItem-root"
      data-rfd-drag-handle-context-id=":ra:"
      data-rfd-drag-handle-draggable-id="3"
      data-rfd-draggable-context-id=":ra:"
      data-rfd-draggable-id="3"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <div>
        child 3
      </div>
    </li>
  </ul>
</div>
`;

exports[`VertListDnd renders with data 1`] = `
<div>
  <ul
    class="MuiList-root MuiList-padding css-134cfso-MuiList-root"
    data-rfd-droppable-context-id=":r3:"
    data-rfd-droppable-id="droppable"
  >
    <li
      aria-describedby="rfd-hidden-text-:r3:-hidden-text-:r4:"
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-6leons-MuiListItem-root"
      data-rfd-drag-handle-context-id=":r3:"
      data-rfd-drag-handle-draggable-id="0"
      data-rfd-draggable-context-id=":r3:"
      data-rfd-draggable-id="0"
      draggable="false"
      role="button"
      tabindex="0"
    >
      item 0
    </li>
    <li
      aria-describedby="rfd-hidden-text-:r3:-hidden-text-:r4:"
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-6leons-MuiListItem-root"
      data-rfd-drag-handle-context-id=":r3:"
      data-rfd-drag-handle-draggable-id="1"
      data-rfd-draggable-context-id=":r3:"
      data-rfd-draggable-id="1"
      draggable="false"
      role="button"
      tabindex="0"
    >
      item 1
    </li>
    <li
      aria-describedby="rfd-hidden-text-:r3:-hidden-text-:r4:"
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-6leons-MuiListItem-root"
      data-rfd-drag-handle-context-id=":r3:"
      data-rfd-drag-handle-draggable-id="2"
      data-rfd-draggable-context-id=":r3:"
      data-rfd-draggable-id="2"
      draggable="false"
      role="button"
      tabindex="0"
    >
      item 2
    </li>
    <li
      aria-describedby="rfd-hidden-text-:r3:-hidden-text-:r4:"
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-6leons-MuiListItem-root"
      data-rfd-drag-handle-context-id=":r3:"
      data-rfd-drag-handle-draggable-id="3"
      data-rfd-draggable-context-id=":r3:"
      data-rfd-draggable-id="3"
      draggable="false"
      role="button"
      tabindex="0"
    >
      item 3
    </li>
  </ul>
</div>
`;
